### ISSUE TITLE
BERT performance on ONNX runtime on Linux CPU much worse than Windows

### ISSUE THREAD

**Issue by micophilip** [microsoft/onnxruntime#4862]
**Describe the bug**
Running inference using bert-base-uncased model on a named entity recognition task using ONNX runtime is significantly worse on Linux than Windows using the exact same model and dataset. The model was fully optimized using `onnxruntime_tools` package

On CentOS Linux 8.2.2004 (Core), Intel Xeon Gold CPU 48-core, it consistently takes **30 minutes** (and 30 minutes using PyTorch 1.5.1)
On Windows Server 2016, Intel Xeon Gold CPU 20-core, it consistently takes **5 minutes** (vs 40 minutes using PyTorch 1.5.1)

Things I tried:
OMP_NUM_THREADS: set to number of cores, set to 1, and unset.
OMP_WAIT_POLICY: ACTIVE and PASSIVE.
intra_op_num_threads: set to 1, 6, number of cores, and unset
fp32 model and fp16

**Urgency**
none

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): CentOS Linux 8.2.2004 (Core)
- ONNX Runtime installed from (source or binary): binary
- ONNX Runtime version: 1.4.0
- Python version: 3.7.8
- Visual Studio version (if applicable): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**To Reproduce**
I followed transformers and ONNX notebook examples. 
``` @python
import torch
# For export:
torch.onnx.export(torch_model,
                  (x, y), 
                  onnx_output,
                  export_params=True,
                  opset_version=11,
                  do_constant_folding=True,
                  input_names=['input_ids', 'token_type_ids'],
                  dynamic_axes={
                  'input_ids': {0: 'batch_size', 1: 'sequence_length'},
                  'token_type_ids': {0: 'batch_size', 1: 'sequence_length'}})

# And for Inference:
provider = 'CPUExecutionProvider'
options = ort.SessionOptions()
options.intra_op_num_threads = 1
session = ort.InferenceSession(onnx_output, options, providers=[provider])
session.run(None, onnx_inputs)
``` 

**Expected behavior**
I expected to see savings in runtime using ONNX vs PyTorch on Linux as it happens on Windows. 

**Screenshots**
N/A

**Additional context**
bert-base-uncased model was finetuned using transformers package v3.0.0 on an NER task.

**Comment by tianleiwu** [comment]
@micophilip, thanks for reporting the issue.

First, please try default session option (NOT set intra_op_num_threads) or set intra_op_num_threads to number of cores. That's the best setting if you use onnxruntime_gpu package.

If that could not solve the issue. Please try [benchmark](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/python/tools/transformers/run_benchmark.sh) on those two machines to see whether it could reproduce the issue.

You can change the settings in the benchmark script like the following (You might update batch size and sequence length according to your data) since your model has no attention mask input and inference is on CPU :
```
run_gpu_fp32=false
run_gpu_fp16=false
run_cpu_fp32=true
run_cpu_int8=false
batch_sizes="1"
sequence_lengths="8 128"
input_counts=2
models_to_test="bert-base-cased"
```

**Comment by micophilip** [comment]
@tianleiwu Thank you for response - Regarding your first comment:

> First, please try default session option (NOT set intra_op_num_threads) or set intra_op_num_threads to number of cores. That's the best setting if you use onnxruntime_gpu package.

Since I've been running inference on CPU, I'm using the `onnxruntime` package and not the GPU one as per https://microsoft.github.io/onnxruntime/. Do you recommend I use `onnxruntime-gpu` on the Linux CPU?

**Comment by tianleiwu** [comment]
@micophilip, 

onnxruntime package is built with OpenMP and onnxruntime_gpu does not. For CPU inference, onnxruntime package usually is better.

**Comment by micophilip** [comment]
@tianleiwu this is the output from result.csv after running the `run_benchmark.sh`, not sure how to interpret this result, but I did notice that the `torchscript` engine test came back on the console much quicker than the `onnxruntime`

|model_name|inputs|engine|version|device|precision|optimizer|io_binding|b1_s8|b1_s128|b1_s512|
|--|--|--|--|--|--|--|--|--|--|--|
|bert-base-cased|1|torchscript|1.5.1|cpu|fp32|||11.13|37.55|109.34|
|bert-base-cased|1|torchscript|1.5.1|cpu|fp32|||10.61|41.44|149.86|
|bert-base-cased|2|onnxruntime|1.4.0|cpu|fp32|True|True|11.28|25.84|90.80|
|bert-base-cased|2|onnxruntime|1.4.0|cpu|fp32|True|False|8.84|28.76|89.44|

**Comment by tianleiwu** [comment]
@micophilip, the table is average latency for different batch size (b) and sequence length (s). It seems that OnnxRuntime is faster in that test. For example batch_size=1 and sequence length 128, latency is 28.76ms for OnnxRuntime, and 37.55ms for PyTorch. That is about 23% speed up.

**Comment by micophilip** [comment]
@tianleiwu I wonder how these results are calculated 🤔 I ran the script again twice, once with only `run_ort=true` and once with only `run_torchscript=true`. In each run, I echoed the timestamp in the `run_benchmark.sh` before the inference and after it. The onnx run took 262 seconds from start to finish of inference, the torchscript took 153 seconds.

Here's the full output with echoed timestamps for the ONNX run
[onnx.log](https://github.com/microsoft/onnxruntime/files/5156975/onnx.log)

Here's the full output with echoed timestamps for the pytorch run
[pytorch.log](https://github.com/microsoft/onnxruntime/files/5156977/pytorch.log)

**Comment by tianleiwu** [comment]
@micophilip, In the benchmark, a random input is generated, and inference latency is measured, and averaged over 1000 times.

ONNX run has more tests than PyTorch (6 vs 3). It is because we tested two settings (with/without IO Binding) in ONNX run.

For example, for input shape [1, 8], ONNX run has two tests (a blank line is added to see it clearly):
```
Tue Sep 1 10:27:14 EDT 2020
Run onnxruntime on bert-base-cased with input shape [1, 8]
{'engine': 'onnxruntime', 'version': '1.4.0', 'device': 'cpu', 'optimizer': True, 'precision': <Precision.FLOAT32: 'fp32'>, 'io_binding': False, 'model_name': 'bert-base-cased', 'inputs': 2, 'batch_size': 1, 'sequence_length': 8, 'datetime': '2020-09-01 10:27:36.371107', 'test_times': 1000, 'latency_variance': '0.00', 'latency_90_percentile': '7.69', 'latency_95_percentile': '7.99', 'latency_99_percentile': '8.55', 'average_latency_ms': '7.50', 'QPS': '133.30'}

Run onnxruntime with io binding on bert-base-cased with input shape [1, 8]
{'engine': 'onnxruntime', 'version': '1.4.0', 'device': 'cpu', 'optimizer': True, 'precision': <Precision.FLOAT32: 'fp32'>, 'io_binding': True, 'model_name': 'bert-base-cased', 'inputs': 2, 'batch_size': 1, 'sequence_length': 8, 'datetime': '2020-09-01 10:27:36.371107', 'test_times': 1000, 'latency_variance': '0.00', 'latency_90_percentile': '7.68', 'latency_95_percentile': '8.42', 'latency_99_percentile': '10.85', 'average_latency_ms': '7.56', 'QPS': '132.28'}
```
PyTorch has only one:
```
Tue Sep 1 10:36:56 EDT 2020
Run PyTorch on bert-base-cased with input shape [1, 8]
{'engine': 'torchscript', 'version': '1.5.1', 'device': 'cpu', 'optimizer': '', 'precision': <Precision.FLOAT32: 'fp32'>, 'io_binding': '', 'model_name': 'bert-base-cased', 'inputs': 1, 'batch_size': 1, 'sequence_length': 8, 'datetime': '2020-09-01 10:37:14.835681', 'test_times': 1000, 'latency_variance': '0.00', 'latency_90_percentile': '10.51', 'latency_95_percentile': '10.83', 
'latency_99_percentile': '11.02', 'average_latency_ms': '10.30', 'QPS': '97.09'}
```
So, if you divide 262 seconds by 2, it will yield 131 seconds. That will be smaller than 153 seconds.

**Comment by micophilip** [comment]
I noticed the benchmark script does not actually fine-tune the BERT model and just runs the inference against the pre-trained model. That's one difference between that and my test case where I actually fine-tune `bert-base-uncased` and run inference on it.

Since none of the recommended options worked for me (such as unsetting `intra_op_num_threads`) and since I'm already using the default settings to `export` and to instantiate an inference session, there's no point of dragging this any longer so I'll just close the issue.

I do encourage the ONNX team however to add to their test suites this scenario of using a fine-tuned `bert-base-uncased` and running inference on it on Linux as it's been consistently slower than PyTorch on Linux CentOS for me no matter what dataset I use.
